lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	) -> Self {
+		Self {
+			params,
+			node_id: NodeId::from_pubkey(node_pubkey),
+			network_graph,
+			channel_liquidities: HashMap::new(),
+		}
+	}
+
+	#[cfg(test)]
+	fn with_channel(mut self, short_channel_id: u64, liquidity: ChannelLiquidity<T>) -> Self {
+		assert!(self.channel_liquidities.insert(short_channel_id, liquidity).is_none());
+		self
+	}
+}
+
+impl Default for ProbabilisticScoringParameters {
+	fn default() -> Self {
+		Self {
+			liquidity_penalty_multiplier_msat: 1000,
+			liquidity_offset_half_life: Duration::from_secs(3600),
+		}
+	}
+}
+
+impl<T: Time> ChannelLiquidity<T> {
+	#[inline]
+	fn new() -> Self {
+		Self {
+			min_liquidity_offset_msat: 0,
+			max_liquidity_offset_msat: 0,
+			last_updated: T::now(),
+		}
+	}
+
+	/// Returns a view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed(
+		&self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&u64, T, &T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&self.min_liquidity_offset_msat, &self.max_liquidity_offset_msat)
+		} else {
+			(&self.max_liquidity_offset_msat, &self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+
+	/// Returns a mutable view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed_mut(
+		&mut self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&mut u64, T, &mut T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&mut self.min_liquidity_offset_msat, &mut self.max_liquidity_offset_msat)
+		} else {
+			(&mut self.max_liquidity_offset_msat, &mut self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &mut self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+}
+
+impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Returns the success probability of routing the given HTLC `amount_msat` through the channel
+	/// in this direction.
+	fn success_probability(&self, amount_msat: u64) -> Probability {
+		let max_liquidity_msat = self.max_liquidity_msat();
+		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
+		if amount_msat > max_liquidity_msat {
+			Probability::Zero
+		} else if amount_msat < min_liquidity_msat {
+			Probability::One
+		} else {
+			let numerator = max_liquidity_msat + 1 - amount_msat;
+			let denominator = max_liquidity_msat + 1 - min_liquidity_msat;
+			if numerator == denominator {
```

Comment:
nit: This will only happen if `amount_msat == min_liquidity_msat` which is why I would change line 648 to: `else if amount_msat <= min_liquidity_msat` and remove this `if` here.

Reply:
I think right now it is 650 and already fixed? might have been the wrong one because of the force push?